### ISSUE TITLE
feat: Support semantic versioning for template versions

### ISSUE THREAD

**Issue by scastria** [copier-org/copier#507]
I would like to make it harder for customers of my template to absorb breaking changes.  The first number in semantic versioning is MAJOR and should be changed for breaking changes.  If I make a breaking change to my template and tag it with a new MAJOR version number, I want the user who runs copier update to have to jump through extra hoops before updating to the breaking change version.

Similar to Terraform, allow the user to specify a version specifier, like 3.x or 3.2.x.  This could be stored in the answers file alongside the _commit value.  So _version_spec could be v3.x and _commit could be v3.1.2.  If I release a new template version as v3.2, then the next time the user runs copier update, copier will know it can upgrade to v3.2 since the user said they wanted any version that starts with 3 MAJOR version number.  If I release a breaking change as v4.0.0, then the next time the user runs copier update, nothing will happen since v4.0.0 doesn't match the user's request version spec of v3.x.  However, copier update could let the user know that a newer version exists.  The user could then run copier update --version-spec v4.x which tells copier to allow the upgrade to v4.0.0 and to update the answers file with the new _version_spec value for the next copier update.

**Comment by Yajo** [comment]
The feature itself makes sense, however I'm more in the side of automating stuff.

ATM you have the `_migrations` option that let you trigger some commands before and after migrating the project to a certain template version. My option has always been to use that to automate migration to higher versions when something wasn't fully backwards-compatible. You could do the same, or you could open some interactive workflow for your specific stuff.

That wouldn't force the user to migrate to a bigger major version only by demand, but would help them having a successful update.

Would that be enough for your use case?

**Comment by scastria** [comment]
I don't think so.  I think it would really be better to allow the user to specify some kind of version spec to control more the upgrades.  Of course, this is just my opinion.  Maybe others can comment.

**Comment by pawamoy** [comment]
I'm tempted to say: sure, why not have the same control over updates as with regular PEP440 specifiers?

`_accept_template_versions: ~=3.2`

**Comment by Yajo** [comment]
fwiw I don't like the idea of handwriting that stuff in the answers file, because that violates the purpose of those `# File generated by copier; DO NOT EDIT MANUALLY` comments that are very helpful.

I think that the cleanest implementation would be to add some extra commands to the CLI/API:

- `copier update-major`
- `copier update-minor`
- `copier update-patch`

However that would kill the purpose of using PEP 440 which is not limited to major.minor.patch as semver. Maybe something like `copier update-compatible 2` which asks for 2 sections compatibility (== minor)? Maybe both, making the 1st 3 ones aliases to the latest one? Maybe a flag in `copier update`? 🤔

**Comment by scastria** [comment]
Agreed. I wasn’t suggesting the user edit the answers file, but to use a new switch on the copier update Command like —version-spec or something similar. 

But I am not too picky on how it is implemented, just as long as the gesture exists. :)

**Comment by Yajo** [comment]
Maybe we can just reuse the `--vcs-ref` option and add support for PEP440 specifiers there.

**Comment by scastria** [comment]
Possibly, but I think the answers file needs to record both the actual version used for the last update and the version specified desired by the user. So if the next time the user wants to update, he can just say copier update without anything else and it will do the right thing. If you make the user always have to specify a vcs-ref during an update, I don’t think it will be good enough as now the user has to remember what version specifier he wants every time.

**Comment by Yajo** [comment]
Well, then it must be recorded indeed, but it must be copier who writes it to the answers file, so we need a flag.

**Comment by shusriva** [comment]
Hi @scastria ,
I am also working on the same use case.
@yajo  Do we have any solution now for this issue ?

**Comment by yajo** [comment]
Sorry, not for now.

The implementation is:

- One new CLI flag: `--version-subscription` with its corresponding API arg
- If passed, check it's a valid PEP 440 specifier.
- Store it in copier answers file under a new key:  `_version_subscription`
- when running copies update, filter incoming tags with it to findthe desired release.

However nobody volunteered yet to implement it.

**Comment by thomaspaulb** [comment]
Thinking about implementing; I would consider changing above spec to:

- Allow user to specify  `_version_subscription` as PEP440 in `copier.yml` rather than a CLI flag, so that new copies of a template get it by default (eg the first release would specify `<2`)
- `copier update` respects it by default
- Add a CLI flag `--ignore-version-subscription` or `--major-upgrade` to ignore it and look for the latest version regardless. (Though maybe not necessary if people can also do this by intentionally specifying `--vcs-ref 2.0.0` to force upgrade from v1 to v2)
- Upon upgrading to v2, `_version_subscription` would be in the template defined as `<3` and the cycle continues

**Comment by pawamoy** [comment]
@thomaspaulb that feels a bit weird to me. It's as if a Python library depended on itself to *force* onto its users the behavior of preventing any major bump by default. My opinion is: just do like any other package manager and let the user decide. We're all responsible adults.

**Comment by thomaspaulb** [comment]
I guess you're right! The only feature we really need for our projects is to be able to pin a legacy project to a certain major version, while still being able to receive updates to that version. I guess that functionality is also covered by the original suggested implementation eg the switch that writes to the copier answers file, after which the project is "pinned". So i will just implement that.

7 feb. 2026 12:05:11 Timothée Mazzucotelli ***@***.***>:
